lighthouse-core/lib/page-functions.js

```diff
@@ -515,34 +518,31 @@ function getNodeDetails(element) {
   return details;
 }
 
-const getNodeDetailsString = `function getNodeDetails(element) {
-  ${getNodePath.toString()};
-  ${getNodeSelector.toString()};
-  ${getBoundingClientRect.toString()};
-  ${getOuterHTMLSnippet.toString()};
-  ${getNodeLabel.toString()};
-  return (${getNodeDetails.toString()})(element);
-}`;
+// HACK: `getNodeDetails` is a function used in many places with multiple deps
+// that would need to be listed in the `evaluate()` call every time it's used.
+// The deps could be inlined, but they need to be available externally for
+// testing. So, override `getNodeDetails.toString()` to include all the deps.
```

Comment:
what this says. All of the functions used in here are only used inside  `getNodeDetails` except `getBoundingClientRect` (which would be easy to replace in `full-page-screenshot` if need be) so they could all be defined internally to a `getNodeDetails` wrapping function but...it is very convenient to test them all separately. Redefining `toString` seems ok to me here, but we could rework if it's too ugly for anyone.

Reply:
this is hilarious